src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/tools/cortisol/IngestTask.java

```diff
@@ -0,0 +1,172 @@
+                    for (int i = 0; i < params.getDocsPerRequest(); ++i) {
+                        String doc = br.readLine();
+                        if (doc == null) {
+                            // restart the indexing from the top.
+                            br = new BufferedReader(new FileReader(params.getTrackLocation()));
+                            doc = br.readLine();
+                        }
+
+                        docs.add(doc);
+                    }
+                    makeBulkRequest(docs);
+                } else {
+                    break;
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private RestHighLevelClient buildEsClient() {
+        final RestClientBuilder builder = RestClient.builder(new HttpHost(endpoint, port, "http"));
+        return new RestHighLevelClient(builder);
+    }
+
+    private void createIndex(final String mappingTemplateJson, int nPrimaryShards, int nReplicaShards) throws IOException {
+        final CreateIndexRequest cir = new CreateIndexRequest(indexName);
+        cir.mapping(mappingTemplateJson, XContentType.JSON);
+        cir.settings(CortisolHelper.buildIndexSettingsJson(nPrimaryShards, nReplicaShards), XContentType.JSON);
+        final CreateIndexResponse response = client.indices().create(cir, RequestOptions.DEFAULT);
+        assert response.isAcknowledged();
+    }
+
+    private void makeBulkRequest(final List<String> bulkDocs) {
+        final BulkRequest bulkRequest = new BulkRequest();
+        for (String bulkDoc : bulkDocs) {
+            final IndexRequest indexRequest = Requests.indexRequest(indexName);
+            indexRequest.source(bulkDoc, XContentType.JSON);
+            bulkRequest.add(indexRequest);
+        }
+        client.bulkAsync(bulkRequest, RequestOptions.DEFAULT, new ActionListener<BulkResponse>() {
+            @Override
+            public void onResponse(BulkResponse bulkItemResponses) {
+                if (!bulkItemResponses.hasFailures()) {
+                    successCount.addAndGet(bulkDocs.size());
+                } else {
+                    for (BulkItemResponse response : bulkItemResponses) {
+                        if (response.isFailed()) {
+                            failureCount.incrementAndGet();
+                        } else {
+                            successCount.incrementAndGet();
+                        }
+                    }
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                failureCount.addAndGet(bulkDocs.size());
+            }
+        });
+        requestCount.addAndGet(bulkDocs.size());
+    }
+
+    private class TokenBucket {
+        private final Lock lock = new ReentrantLock();
+        private final Condition hasTokens = lock.newCondition();
+        private final ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
```

Comment:
I think you can avoid having a scheduler here. Have calls to take() block, and make sure every take() refills what it took from the bucket once it's complete.

Reply:
take() already blocks with await() but I see what you mean here. The idea is to make sure that we make a fixed number of calls every second. If we made the call to makeBulkRequest() block and replenish the tokens once the requests are done, then we won't be able to guarantee a constant request rate. We don't want to replenish too soon(if bulk calls finish faster) or too late(if the bulk calls take longer to finish).